java/ddlogapi.c

```diff
@@ -260,6 +271,40 @@ JNIEXPORT jint JNICALL Java_ddlogapi_DDlogAPI_ddlog_1enable_1cpu_1profiling(
     return ddlog_enable_cpu_profiling((ddlog_prog)progHandle, enable);
 }
 
+void log_callback(uintptr_t callbackInfo, int level, const char *msg) {
+    struct CallbackInfo* cbi = (struct CallbackInfo*)callbackInfo;
+    JNIEnv* env;
+    (*cbi->jvm)->AttachCurrentThreadAsDaemon(cbi->jvm, (void**)&env, NULL);
+    assert(env);
+    jstring jmsg = (*env)->NewStringUTF(env, msg);
+    (*env)->CallVoidMethod(env, cbi->obj, cbi->method, jmsg, (jint) level);
+}
+
+/* This function sets up a new logging callback for the given module and
+ * then deallocates the old `CallbackInfo`, if any.
```

Comment:
You should document the return value; it's a handle to a callbackinfo data structure.
A return of 0 seems to be possible even in case of success - when you just wanted to delete the callback.